contentful/errors.py

```diff
@@ -17,58 +20,141 @@ class HTTPError(Exception):
+        response_json = None
+        error_message = [
+          "HTTP status code: {0}".format(self.status_code),
+        ]
+        try:
+            response_json = response.json()
+
+            message = response_json.get('message', None)
+            details = response_json.get('details', None)
+            request_id = response_json.get('requestId', None)
+
+            if message is not None:
+                error_message.append("Message: {0}".format(message))
+            else:
+                error_message.append("Message: {0}".format(self._default_error_message()))
+            if details is not None:
+                error_message.append("Details: {0}".format(self._handle_details(details)))
+            if request_id is not None:
+                error_message.append("Request ID: {0}".format(request_id))
+        except json.JSONDecodeError:
+            error_message.append("Message: {0}".format(self._default_error_message()))
+        finally:
+            return "\n".join(error_message)
 
 
 class BadRequestError(HTTPError):
     """
     400
     """
-    pass
+
+    def _default_error_message(self):
+        return "The request was malformed or missing a required parameter."
+
+    def _handle_details(self, details):
+        from .utils import string_class
+        if isinstance(details, string_class()):
+            return details
+
+        def _handle_detail(detail):
+            if isinstance(detail, string_class()):
+                return detail
+            return detail.get('details', None)
+
+        inner_details = [_handle_detail(detail) for detail in details['errors']]
+        inner_details = [detail for detail in inner_details if detail is not None] # This works in both Py2 and Py3
+        return "\n\t".join(inner_details)
+
+
+class UnauthorizedError(HTTPError):
+    """
+    401
+    """
+
+    def _default_error_message(self):
+        return "The authorization token was invalid."
 
 
 class AccessDeniedError(HTTPError):
     """
     403
     """
-    pass
 
+    def _default_error_message(self):
+        return "The specified token does not have access to the requested resource."
 
-class UnauthorizedError(HTTPError):
+
+    def _handle_details(self, details):
+        return "\n\tReasons:\n\t\t{0}".format("\n\t\t".join(details['reasons']))
+
+
+class NotFoundError(HTTPError):
     """
-    401
+    404
     """
-    pass
+
+    def _default_error_message(self):
+        return "The requested resource or endpoint could not be found."
+
+    def _handle_details(self, details):
+        message = "The requested {0} could not be found.".format(details['type'])
+        resource_id = details.get('id', None)
+        if resource_id is not None:
+            message += " ID: {0}.".format(resource_id)
+
+        return message
 
 
 class RateLimitExceededError(HTTPError):
     """
     429
     """
-    pass
+
+    def _default_error_message(self):
+        return "Rate limit exceeded. Too many requests."
```

Comment:
Question for context: Can you parse the http headers here to include the time the user have to wait, or the current rate limit headers states?

Reply:
Yes, I can do that.

I'll add it.

Reply:
Done